esmvaltool/diag_scripts/ipcc_ar5/ch09_fig09_17.py

```diff
@@ -0,0 +1,431 @@
+"""
+Time series diagnostics
+=======================
+
+Diagnostic to produce figures of the time development of a field from
+cubes. These plost show time on the x-axis and cube value (ie temperature) on
+the y-axis.
+
+Two types of plots are produced: individual model timeseries plots and
+multi model time series plots. The inidivual plots show the results from a
+single cube, even if this is a mutli-model mean made by the _multimodel.py
+preproccessor. The multi model time series plots show several models
+on the same axes, where each model is represented by a different line colour.
+
+Note that this diagnostic assumes that the preprocessors do the bulk of the
+hard work, and that the cube received by this diagnostic (via the settings.yml
+and metadata.yml files) has a time component, no depth component, and no
+latitude or longitude coordinates.
+
+An approproate preprocessor for a 3D+time field would be::
+
+  preprocessors:
+    prep_timeseries_1:# For Global Volume Averaged
+      average_volume:
+        coord1: longitude
+        coord2: latitude
+        coordz: depth
+
+An approproate preprocessor for a 3D+time field at the surface would be::
+
+    prep_timeseries_2: # For Global surface Averaged
+      extract_levels:
+        levels:  [0., ]
+        scheme: linear_extrap
+      average_area:
+        coord1: longitude
+        coord2: latitude
+
+An approproate preprocessor for a 2D+time field would be::
+
+    prep_timeseries_2: # For Global surface Averaged
+      average_area:
+        coord1: longitude
+        coord2: latitude
+
+This tool is part of the ocean diagnostic tools package in the ESMValTool.
+
+Author: Lee de Mora (PML)
+        ledm@pml.ac.uk
+"""
+
+import logging
+import os
+
+import iris
+import matplotlib.pyplot as plt
+import numpy as np
+
+from esmvaltool.diag_scripts.ocean import diagnostic_tools as diagtools
+from esmvaltool.diag_scripts.shared import run_diagnostic
+
+# This part sends debug statements to stdout
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+def timeplot(cube, **kwargs):
+    """
+    Create a time series plot from the cube.
+
+    Note that this function simple does the plotting, it does not save the
+    image or do any of the complex work. This function also takes and of the
+    key word arguments accepted by the matplotlib.pyplot.plot function.
+    These arguments are typically, color, linewidth, linestyle, etc...
+
+    If there's only one datapoint in the cube, it is plotted as a
+    horizontal line.
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+        Input cube
+
+    """
+    cubedata = np.ma.array(cube.data)
+    if len(cubedata.compressed()) == 1:
+        plt.axhline(cubedata.compressed(), **kwargs)
+        return
+
+    times = diagtools.cube_time_to_float(cube)
+    plt.plot(times, cubedata, **kwargs)
+
+
+def moving_average(cube, window):
```

Comment:
@tillku this function can be replaced (and optimized thereafter) by iris's builtin `rolling_window`: https://scitools.org.uk/iris/docs/v2.0/iris/iris/cube.html#iris.cube.Cube.rolling_window - you could define it as say `rolling_window(cube.coord('time'), iris.analysis.MEAN, roll_win, **kwargs)` where you define `roll_win` according to your needs; this would eliminate the unnecessary masking and even worse (from a memory usage point of view) building a list of arrays (which you should never do since lists are very heavy on memory in Python especially if their members are full-blown arrays)

Reply:
@valeriupredoi I think this function is not being used actually, I've commented it our for now.